### DISCUSSION TITLE
Customize shared template

### DISCUSSION THREAD

**Discussion by jonocodes** [aminalaee/sqladmin#762] [General]
I would like to add some javascript and css that will affect all the pages.

For now I am doing that individually on each custom template. For example:

organization_edit.html
```
{% extends "edit.html" %}

{% block tail %}
  <script type="text/javascript">
      $(document).ready(function() {
          $('select[multiple]').select2()
      });
  </script>
{% endblock %}
```

I would like to make this change either to the layout.html file to affect the whole site. Or do it in a more targeted fashion to the 'edit.html' template so all edit pages affected.

**Comment by hasansezertasan** [comment]
You can simply override [base.html](https://github.com/aminalaee/sqladmin/blob/main/sqladmin%2Ftemplates%2Fsqladmin%2Fbase.html).

Flask Admin has some examples of doing this in the repo. Here are the four examples I could find:

- [flask-admin/examples/simple at master · flask-admin/flask-admin](https://github.com/flask-admin/flask-admin/tree/master/examples/simple)
- [flask-admin/examples/custom-layout at master · flask-admin/flask-admin](https://github.com/flask-admin/flask-admin/tree/master/examples/custom-layout)
- [flask-admin/examples/auth at master · flask-admin/flask-admin](https://github.com/flask-admin/flask-admin/tree/master/examples/auth)
- [flask-admin/examples/auth-flask-login at master · flask-admin/flask-admin](https://github.com/flask-admin/flask-admin/tree/master/examples/auth-flask-login)

I think we can add a recipe to the cookbook that demonstrates this but an examples folder could be better, WDYT @aminalaee?

**Reply by hasansezertasan** [reply]
Visiting this again, I'm down to work on some examples like in the Flask Admin repo but I want to know what @aminalaee thinks about it.

The examples folder in the Flask Admin repo did (and still does) helped me a lot when creating a project with it.

**Comment by jonocodes** [comment]
Some other ideas for implementing this feature.

1. It appears templates are now referenced to in the 'sqladmin' dir. For example:
	`{% extends "sqladmin/layout.html" %}`
	
	Going forward, if the user specifies:
	`{% extends "layout.html" %}`

	Then it can look for 'layout.html' in templates_dir and use that.

	I guess that roughly means any template that does not start with sqladmin/ will be looked for in templates_dir regardless of the name. For example:
	`{% extends "my/great/layout.html" %}`


2. Search the templates_dir for templates with the same name before looking in the sqladmin package.
So specifying this:
	`{% extends "sqladmin/layout.html" %}`

	Will first look in templates_dir and use it if found. It not found, use the library template.

**Reply by hasansezertasan** [reply]
Flask Admin has a great solution for this.

Checkout these links and follow the `admin_base_template` and `base_template` parameters.

- https://github.com/flask-admin/flask-admin/blob/14e24c970f0ee3a29add830612eee9a0b0ba5dcc/flask_admin/templates/bootstrap4/admin/master.html#L1C12-L1C31
- https://github.com/flask-admin/flask-admin/blob/14e24c970f0ee3a29add830612eee9a0b0ba5dcc/flask_admin/base.py#L468

@aminalaee, what do you think about Flask Admin's solution on this? Do we need such an approach?

**Reply by hasansezertasan** [reply]
Also, the user will know what to specify because the documentation is updated accordingly and there are only two resources for the user to know how to override templates: source code and documentation.

**Comment by jonocodes** [comment]
> You can simply override [base.html](https://github.com/aminalaee/sqladmin/blob/main/sqladmin%2Ftemplates%2Fsqladmin%2Fbase.html).

Thanks @hasansezertasan . Is this a recommendation to me as a user, or is this an idea for a new feature in sqladmin.

If the first how would I "override" base.html? Some kind of package monkey patching? I'm not sure how to do that.

**Reply by hasansezertasan** [reply]
> Thanks @hasansezertasan . Is this a recommendation to me as a user, or is this an idea for a new feature in sqladmin.

You're welcome. It's for you as a user. 🤓

> If the first how would I "override" base.html? Some kind of package monkey patching? I'm not sure how to do that.

No, that's a terrible idea.

Here are steps to override:

- Create a `templates` folder.
- If you are using the latest release (0.17.0), create `sqladmin` folder in the `templates` folder, otherwise skip this step.
- Copy the `base.html` into the folder you created in the previous step.
- Update `base.html` accordingly.

My example:

```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover" />
  <link rel="stylesheet" href="{{ url_for('admin:statics', path='css/tabler.min.css') }}">
  <link rel="stylesheet" href="{{ url_for('admin:statics', path='css/fontawesome.min.css') }}">
  <link rel="stylesheet" href="{{ url_for('admin:statics', path='css/select2.min.css') }}">
  <link rel="stylesheet" href="{{ url_for('admin:statics', path='css/flatpickr.min.css') }}">
  <link rel="stylesheet" href="{{ url_for('admin:statics', path='css/main.css') }}">
  {% block head %}
  {% endblock %}
  <title>{{ admin.title }}</title>
</head>

<body>
  {% block body %}
  <main>
    {% block main %}
    {% endblock %}
  </main>
  {% endblock %}
  <script type="text/javascript" src="{{ url_for('admin:statics', path='js/jquery.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('admin:statics', path='js/tabler.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('admin:statics', path='js/popper.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('admin:statics', path='js/bootstrap.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('admin:statics', path='js/select2.full.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('admin:statics', path='js/flatpickr.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('admin:statics', path='js/main.js') }}"></script>
  <script>
    console.log("I'm everywhere!")
  </script>
  {% block tail %}
  {% endblock %}
</body>

</html>
```

If you use the example above, you'll see "I'm everywhere!" text logged in the browser console, on all pages.

Let me know if you encounter any problems.

**Reply by jonocodes** [reply]
I'm on sqladmin 0.16.0 and tried this. No dice.
I created a layout.html and base.html file in my templates dir. And they were not picked up when the template was rendered. It would be awesome it worked.
I know that dir works since I am sourcing in other custom templates from there like user_edit.html

**Reply by jonocodes** [reply]
Oh, I take it back. That worked. Awesome!
This should be documented since its super useful.

**Reply by aminalaee** [reply]
Feel free to document it as you think fit.

**Reply by jonocodes** [reply]
Well after all that when I went to update docs/working_with_templates.md I see that this was covered right at the top of the doc. So I just expanded it a bit:
https://github.com/aminalaee/sqladmin/pull/769

**Reply by hasansezertasan** [reply]
> Oh, I take it back. That worked. Awesome!
> This should be documented since its super useful.

Happy to help 👍.

**Reply by hasansezertasan** [reply]
> Well after all that when I went to update docs/working_with_templates.md I see that this was covered right at the top of the doc. So I just expanded it a bit:
> https://github.com/aminalaee/sqladmin/pull/769
> 

I think the "copy and paste" approach is terrible (yep, I was the one who came up with that solution to save the day). @aminalaee would know, Flask Admin has parameters called `admin_base_template` and `base_template` to make overriding/extending process easier, I have mentioned above at https://github.com/aminalaee/sqladmin/discussions/762#discussioncomment-9427086.

Check out the [Flask Admin Documentation](https://flask-admin.readthedocs.io/en/latest/introduction/#extending-the-built-in-templates).

We need that...

**Reply by jonocodes** [reply]
Yeah, if you could extend instead of overwrite that would be great.
